### PR TITLE
shadow service

### PR DESCRIPTION
**Pull request by zouyingjie**
1. emctl: add CURD processing for ShadowServices.
2. shadow service controller:  Create a copy of the service's Deployment or MeshDeployment based on shadow service.

### CODE REVIEW

**Comment by xxx7xxxx** [comment]
@zouyingjie You need to fix the conflict

**Review by zhao-kun — CHANGES_REQUESTED** [review]
1. All source files haven't license declaration
2. Too many COPY codes, please remove them (you could introduce from original pkg via import ....) . **PLEASE DO NOT COPY/PAST CODE**
3.  Please fix complaints from the Github Action

File: shadow/pkg/handler/deployment_cloner.go [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+			return errors.Wrapf(err, "Clone deployment %s for service %s failed", sourceDeployment.Name, shadowService.ServiceName)
+		}
+		return err
+	}
+}
+
+func (cloner *ShadowServiceCloner) cloneDeploymentSpec(sourceDeployment *appsV1.Deployment, shadowService *object.ShadowService) *appsV1.Deployment {
+	shadowDeployment := cloner.injectShadowConfiguration(
+		cloner.shadowDeploymentBaseSpec(
+			cloner.shadowDeploymentInitialize(nil)))(sourceDeployment, shadowService)
```
Comment (localvar) [inline shadow/pkg/handler/deployment_cloner.go:51]: the long code line is hard to read, propose to split to short statements.
and it seems we don't need to use closure functions, they make code hard to read.

File: shadow/pkg/handler/deployment_cloner.go [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+	}
+
+	env := &corev1.EnvVar{}
+	env.Name = envName
+	env.Value = string(configJson)
+	return env
+
+}
+
+func (cloner *ShadowServiceCloner) shadowDeploymentInitialize(fn cloneDeploymentSpecFunc) cloneDeploymentSpecFunc {
```
Comment (localvar) [inline shadow/pkg/handler/deployment_cloner.go:154]: parameter `fn` is not used.

**Review by localvar — APPROVED** [review]

**Review by zhao-kun — CHANGES_REQUESTED** [review]
Please provide unit tests, and make sure the coverage reaches 80% @zouyingjie

File: shadow/pkg/controller/shadowservice_controller.go [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+
+	kubernetesClient, err := utils.NewKubernetesClient()
+	if err != nil {
+		return nil, errors.Wrapf(err, "new kubernetes clientSet error")
+	}
+	runtimeClient, err := utils.NewRuntimeClient()
+	if err != nil {
+		return nil, errors.Wrapf(err, "new Controller Runtime client error")
+	}
+	// crdRestClient, err := utils.NewCRDRestClient()
```
[the remark this answers is not in the document]
Comment (zouyingjie) [reply shadow/pkg/controller/shadowservice_controller.go]: Updated

File: shadow/pkg/controller/shadowservice_controller.go [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+		RunTimeClient: &runtimeClient,
+		DeleteChan:    deleteChan,
+	}
+
+	shadowServiceSyncer, err := syncer.NewSyncer(config.MeshServer, config.RequestTimeout, config.PullInterval)
+	return &ShadowServiceController{kubernetesClient, &runtimeClient, shadowServiceSyncer,
+		&shadowServiceCloner, &shadowServiceSearcher, &shadowServiceSearcherDeleter, cloneChan, deleteChan}, nil
+}
+
+// Do start shadow service sync and clone.
```
[the remark this answers is not in the document]
Comment (zouyingjie) [reply shadow/pkg/controller/shadowservice_controller.go]: Updated

File: shadow/pkg/handler/deployment_cloner.go [hunk 5]
```diff
@@ -0,0 +1,275 @@
...
+	shadowDeployment := cloner.generateShadowDeployment(sourceDeployment, shadowService)
+	cloner.decorateShadowDeploymentBaseSpec(shadowDeployment, sourceDeployment)
+	cloner.decorateShadowConfiguration(shadowDeployment, sourceDeployment, shadowService)
+	return shadowDeployment
+}
+
+// findContainer returns the copy of the container,
+// which means it won't change the original container when changing the result.
+func findContainer(containers []corev1.Container, containerName string) (*corev1.Container, bool) {
+
```
[the remark this answers is not in the document]
Comment (zouyingjie) [reply shadow/pkg/handler/deployment_cloner.go]: Updated

File: shadow/pkg/object/spec.go [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package object
+
+type (
+	MySQL struct {
```
[the remark this answers is not in the document]
Comment (zouyingjie) [reply shadow/pkg/object/spec.go:21]: Updated

**Review by xxx7xxxx — APPROVED** [review]

File: shadow/pkg/controller/shadowservice_controller.go [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+	"github.com/megaease/easemesh/mesh-shadow/pkg/syncer"
+	"github.com/megaease/easemesh/mesh-shadow/pkg/utils"
+
+	"github.com/pkg/errors"
+	"k8s.io/client-go/kubernetes"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	ShadowServiceKind = "ShadowService"
```
Comment (zhao-kun) [inline shadow/pkg/controller/shadowservice_controller.go:34]: Fix complains

File: shadow/pkg/handler/clone_handler.go [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+	shadowDeploymentNameSuffix = "-shadow"
+
+	databaseShadowConfigEnv      = "EASE_RESOURCE_DATABASE"
+	kafkaShadowConfigEnv         = "EASE_RESOURCE_KAFKA"
+	rabbitmqShadowConfigEnv      = "EASE_RESOURCE_RABBITMQ"
+	redisShadowConfigEnv         = "EASE_RESOURCE_REDIS"
+	elasticsearchShadowConfigEnv = "EASE_RESOURCE_ELASTICSEARCH"
+)
+
+type Cloner interface {
```
Comment (zhao-kun) [inline shadow/pkg/handler/clone_handler.go:55]: ditto

File: shadow/pkg/handler/delete_handler.go [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+	"github.com/megaease/easemesh/mesh-shadow/pkg/object"
+	"github.com/megaease/easemesh/mesh-shadow/pkg/utils"
+	appv1 "k8s.io/api/apps/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	runTimeClient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	Separator = '/'
```
Comment (zhao-kun) [inline shadow/pkg/handler/delete_handler.go:32]: ditto

File: shadow/pkg/handler/deployment_cloner.go [hunk 10]
```diff
@@ -0,0 +1,274 @@
...
+	}
+	return envVars
+}
+
+func generateShadowConfigEnv(envName string, config interface{}) *corev1.EnvVar {
+	if config == nil || reflect.ValueOf(config).IsNil() {
+		return nil
+	}
+
+	configJson, err := json.Marshal(config)
```
Comment (zhao-kun) [inline shadow/pkg/handler/deployment_cloner.go:114]: ditto

File: shadow/pkg/handler/search_handler.go [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+	"github.com/megaease/easemesh/mesh-shadow/pkg/object"
+	"github.com/megaease/easemesh/mesh-shadow/pkg/utils"
+	appsV1 "k8s.io/api/apps/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	runTimeClient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	MeshServiceAnnotation = "mesh.megaease.com/service-name"
```
Comment (zhao-kun) [inline shadow/pkg/handler/search_handler.go:32]: ditto

File: shadow/pkg/object/spec.go [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+		Uris     string `yaml:"uris" json:"uris" jsonschema:"required"`
+		UserName string `yaml:"userName" json:"userName" jsonschema:"required"`
+		Password string `yaml:"password" json:"password" jsonschema:"required"`
+	}
+
+	Kafka struct {
```
Comment (zhao-kun) [inline shadow/pkg/object/spec.go:27]: ditto

File: shadow/pkg/object/spec.go [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+		Uris string `yaml:"uris" json:"uris" jsonschema:"required"`
+	}
+
+	Redis struct {
```
Comment (zhao-kun) [inline shadow/pkg/object/spec.go:31]: ditto

Comment (zouyingjie) [reply shadow/pkg/object/spec.go:34]: updated

File: shadow/pkg/object/spec.go (hunk 12 above)
Comment (zouyingjie) [reply shadow/pkg/object/spec.go:29]: updated

File: shadow/pkg/controller/shadowservice_controller.go (hunk 7 above)
Comment (zouyingjie) [reply shadow/pkg/controller/shadowservice_controller.go]: updated

**Review by zhao-kun — APPROVED** [review]

**Pull request merged**